packages/editor/src/components/rich-text/editable.js

```diff
@@ -0,0 +1,199 @@
+ * WordPress dependencies
+ */
+import { Component, createElement } from '@wordpress/element';
+import { BACKSPACE, DELETE } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import { diffAriaProps, pickAriaProps } from './aria';
+
+/**
+ * Browser dependencies
+ */
+
+const { userAgent } = window.navigator;
+
+/**
+ * Applies a fix that provides `input` events for contenteditable in Internet Explorer.
+ *
+ * @param {Element} editorNode The root editor node.
+ *
+ * @return {Function} A function to remove the fix (for cleanup).
+ */
+function applyInternetExplorerInputFix( editorNode ) {
+	/**
+	 * Dispatches `input` events in response to `textinput` events.
+	 *
+	 * IE provides a `textinput` event that is similar to an `input` event,
+	 * and we use it to manually dispatch an `input` event.
+	 * `textinput` is dispatched for text entry but for not deletions.
+	 *
+	 * @param {Event} textInputEvent An Internet Explorer `textinput` event.
+	 */
+	function mapTextInputEvent( textInputEvent ) {
+		textInputEvent.stopImmediatePropagation();
+
+		const inputEvent = document.createEvent( 'Event' );
+		inputEvent.initEvent( 'input', true, false );
+		inputEvent.data = textInputEvent.data;
+		textInputEvent.target.dispatchEvent( inputEvent );
+	}
+
+	/**
+	 * Dispatches `input` events in response to Delete and Backspace keyup.
+	 *
+	 * It would be better dispatch an `input` event after each deleting
+	 * `keydown` because the DOM is updated after each, but it is challenging
+	 * to determine the right time to dispatch `input` since propagation of
+	 * `keydown` can be stopped at any point.
+	 *
+	 * It's easier to listen for `keyup` in the capture phase and dispatch
+	 * `input` before `keyup` propagates further. It's not perfect, but should
+	 * be good enough.
+	 *
+	 * @param {KeyboardEvent} keyUp
+	 * @param {Node}          keyUp.target  The event target.
+	 * @param {number}        keyUp.keyCode The key code.
+	 */
+	function mapDeletionKeyUpEvents( { target, keyCode } ) {
+		const isDeletion = BACKSPACE === keyCode || DELETE === keyCode;
+
+		if ( isDeletion && editorNode.contains( target ) ) {
+			const inputEvent = document.createEvent( 'Event' );
+			inputEvent.initEvent( 'input', true, false );
+			inputEvent.data = null;
+			target.dispatchEvent( inputEvent );
+		}
+	}
+
+	editorNode.addEventListener( 'textinput', mapTextInputEvent );
+	document.addEventListener( 'keyup', mapDeletionKeyUpEvents, true );
+	return function removeInternetExplorerInputFix() {
+		editorNode.removeEventListener( 'textinput', mapTextInputEvent );
+		document.removeEventListener( 'keyup', mapDeletionKeyUpEvents, true );
+	};
+}
+
+const IS_PLACEHOLDER_VISIBLE_ATTR_NAME = 'data-is-placeholder-visible';
+const CLASS_NAME = 'editor-rich-text__editable';
+
+/**
+ * Whether or not the user agent is Internet Explorer.
+ *
+ * @type {boolean}
+ */
+const IS_IE = userAgent.indexOf( 'Trident' ) >= 0;
+
+export default class Editable extends Component {
```

Comment:
Having a hard time understanding how do we decide what belongs in this class which I believe is the replacement for the TinyMCE component and what belongs in RichText. The line is very blury right now. I do see a need for splitting to keep the code contained. 

It doesn't need an immediate resolution but maybe there are better ways to split the RichText component.

Reply:
I see this component as just an enhanced `contenteditable` element, that's why I called it editable. The main things it does is preventing React from rerendering it and applying some browser specific fixes.

Reply:
One of my next PRs will be to use React to render the children of this component. Then we can get rid of the custom attribute diffing and `shouldComponentUpdate`.

Reply:
So eventually I think this class will be reduced to nothing.